pkg/controller/cloudprofile/cloudprofile_control.go

```diff
@@ -0,0 +1,171 @@
+	c.cloudProfileAdd(newObj)
+}
+
+func (c *Controller) cloudProfileDelete(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.cloudProfileQueue.Add(key)
+}
+
+func (c *Controller) reconcileCloudProfileKey(key string) error {
+	// Get the cloudprofile to the key from cache
+	_, cloudProfileName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	cloudProfile, err := c.cloudProfileLister.Get(cloudProfileName)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Debugf("[CLOUDPROFILE RECONCILE] %s - skipping because CloudProfile has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		logger.Logger.Infof("[CLOUDPROFILE RECONCILE] %s - unable to retrieve object from store: %v", key, err)
+		return err
+	}
+
+	err = c.control.ReconcileCloudProfile(cloudProfile, key)
+	if err != nil {
+		c.cloudProfileQueue.AddAfter(key, 15*time.Second)
+	}
+	return nil
+}
+
+// ControlInterface implements the control logic for reconciling CloudProfiles. It is implemented as an interface to allow
+// for extensions that provide different semantics. Currently, there is only one implementation.
+type ControlInterface interface {
+	// ReconcileCloudProfile implements the control logic for CloudProfile creation, update, and deletion.
+	// If an implementation returns a non-nil error, the invocation will be retried using a rate-limited strategy.
+	// Implementors should sink any errors that they do not wish to trigger a retry, and they may feel free to
+	// exit exceptionally at any point provided they wish the update to be re-run at a later point in time.
+	ReconcileCloudProfile(cloudprofile *gardenv1beta1.CloudProfile, key string) error
+}
+
+// NewDefaultControl returns a new instance of the default implementation ControlInterface that
+// implements the documented semantics for CloudProfiles.
+func NewDefaultControl(k8sGardenClient kubernetes.Client, seedLister gardenlisters.SeedLister, shootLister gardenlisters.ShootLister) ControlInterface {
+	return &defaultControl{k8sGardenClient, seedLister, shootLister}
+
+}
+
+type defaultControl struct {
+	k8sGardenClient kubernetes.Client
+	seedLister      gardenlisters.SeedLister
+	shootLister     gardenlisters.ShootLister
+}
+
+func (c *defaultControl) ReconcileCloudProfile(obj *gardenv1beta1.CloudProfile, key string) error {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return err
+	}
+
+	var (
+		cloudProfile       = obj.DeepCopy()
+		cloudProfileLogger = logger.NewCloudProfileLogger(logger.Logger, cloudProfile.Name)
+	)
+
+	// The deletionTimestamp labels the CloudProfile as intended to get deleted. Before deletion, it has to be ensured that
+	// no Shoots and Seed s are assigned to the CloudProfile anymore. If this is the case then the controlller will remove
+	// the Finalizers from the CloudProfile and deletion will be triggered.
```

Comment:
Finalizers -> finalizers
"and deletion will be triggered" -> "so that it can be garbage collected"

Reply:
done